src/test/java/uk/gov/pay/webhooks/deliveryqueue/managed/WebhookMessagePollingServiceTest.java

```diff
@@ -0,0 +1,138 @@
+        instantSource = InstantSource.fixed(Instant.now());
+        webhookMessageDao = new WebhookMessageDao(database.getSessionFactory());
+        webhookDao = new WebhookDao(database.getSessionFactory());
+        webhookDeliveryQueueDao = new WebhookDeliveryQueueDao(database.getSessionFactory(), instantSource);
+
+        sendAttempter = new SendAttempter(webhookDeliveryQueueDao, instantSource, webhookMessageSenderMock, environment);
+        webhookMessagePollingService = new WebhookMessagePollingService(webhookDeliveryQueueDao, sendAttempter, database.getSessionFactory());
+    }
+
+    @Test
+    public void shouldPollWithNoActionsOnEmptyQueue() {
+       webhookMessagePollingService.pollWebhookMessageQueue();
+       verifyNoInteractions(webhookMessageSenderMock);
+    }
+
+    @Test
+    public void shouldEmitSingleValidQueueItem() throws IOException, InvalidKeyException, InterruptedException {
+        setupOrderedDeliveryQueueWith(List.of("first-external-id"));
+        webhookMessagePollingService.pollWebhookMessageQueue();
+        verify(webhookMessageSenderMock).sendWebhookMessage(argThat((webhookMessage -> webhookMessage.getExternalId().equals("first-external-id"))));
+    }
+
+    @Test
+    public void shouldEmitMultipleValidQueueItemsInOnePoll() throws IOException, InvalidKeyException, InterruptedException {
+        setupOrderedDeliveryQueueWith(List.of("first-external-id", "second-external-id"));
+        webhookMessagePollingService.pollWebhookMessageQueue();
+
+        var captor = ArgumentCaptor.forClass(WebhookMessageEntity.class);
+        verify(webhookMessageSenderMock, times(2)).sendWebhookMessage(captor.capture());
+        assertEquals("first-external-id", captor.getAllValues().get(0).getExternalId());
+        assertEquals("second-external-id", captor.getAllValues().get(1).getExternalId());
+    }
+
+    @Test
+    public void shouldAppropriatelyHandleFailedEmitAndNotTryAgainImmediately() throws IOException, InvalidKeyException, InterruptedException {
+        when(response.statusCode()).thenReturn(500);
+
+        setupOrderedDeliveryQueueWith(List.of("first-external-id"));
+        webhookMessagePollingService.pollWebhookMessageQueue();
+        verify(webhookMessageSenderMock).sendWebhookMessage(argThat((webhookMessage -> webhookMessage.getExternalId().equals("first-external-id"))));
+
+        webhookMessagePollingService.pollWebhookMessageQueue();
+        verifyNoMoreInteractions(webhookMessageSenderMock);
+    }
+
+    private void setupOrderedDeliveryQueueWith(List<String> messageIds) {
+        WebhookEntity webhookEntity = new WebhookEntity();
+
+        database.inTransaction(() -> {
+            webhookEntity.setCallbackUrl("http://a-callback-url.com");
+            webhookEntity.setServiceId("a-service-id");
+            webhookEntity.setLive(false);
+
+            webhookDao.create(webhookEntity);
+        });
+
+        for (int i = 0; i < messageIds.size(); i++) {
+            int finalI = i;
```

Comment:
It took me a while to realise `final` is an effectively final variable, rather than the last value `i` gets or something.

You _could_ do this:

```java
IntStream.range(0, messageIds.size()).forEach(i -> {
    var message = database.inTransaction(() -> {
        WebhookMessageEntity webhookMessageEntity = new WebhookMessageEntity();
        var id = messageIds.get(i);
        // …
        });
    // …
});
```